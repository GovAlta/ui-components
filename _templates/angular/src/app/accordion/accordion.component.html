<h2>Accordion</h2>

<goab-accordion heading="This is the title">
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Animi tempora optio illum
    iste consectetur harum vel voluptatem eos quos deleniti, sequi sapiente porro culpa
    repellendus eligendi in architecto natus omnis. Lorem ipsum dolor sit amet consectetur
    adipisicing elit. Repudiandae, nam corporis. Omnis tempora non aspernatur nostrum
    pariatur quibusdam magni eum vitae similique, laboriosam reiciendis repellat sapiente.
    Nobis laudantium facilis incidunt!
  </p>
  <goa-button (_click)="onClick()" size="compact" type="primary">Link to application</goa-button>
</goab-accordion>

<goab-accordion heading="This is the title" secondarytext="Secondary text">
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Quidem sapiente delectus, hic
    explicabo doloribus magnam, consequatur eaque veritatis possimus porro ducimus vero
    ipsa omnis, soluta sit voluptates quam facilis odio. Lorem, ipsum dolor sit amet
    consectetur adipisicing elit. Animi, officia magnam assumenda nisi exercitationem
    rerum blanditiis natus modi nobis error iure asperiores? Iste vero odio magnam,
    reprehenderit doloribus error iusto.
  </p>
</goab-accordion>

<goab-button type="tertiary" mb="xl" (_click)="onClick()">
  Toggle Accordion Group
</goab-button>

<goabx-button type="tertiary" mb="xl" (_click)="onClick()">
  This is the GoABX
</goabx-button>

<goab-accordion heading="Heading" [open]="open"> Content 1 </goab-accordion>
<goab-accordion heading="Heading" [open]="open">
  <goab-badge type="success" content="Success" slot="headingcontent"></goab-badge>
  Content 2
</goab-accordion>
<goab-accordion heading="Heading" [open]="open"> Content 3 </goab-accordion>
