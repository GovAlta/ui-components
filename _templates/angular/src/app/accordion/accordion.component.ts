import { GoabAccordion, GoabBadge, GoabButton } from "@abgov/angular-components";
import { CUSTOM_ELEMENTS_SCHEMA, Component } from "@angular/core";

@Component({
  standalone: true,
  selector: "abgov-accordion",
  templateUrl: "./accordion.component.html",
  schemas: [CUSTOM_ELEMENTS_SCHEMA],
  imports: [
    GoabAccordion,
    GoabButton,
    GoabBadge,
  ],
})
export class AccordionComponent {
  constructor() { }
  open = false;
  onClick() {
    this.open = !this.open;
  }
}
