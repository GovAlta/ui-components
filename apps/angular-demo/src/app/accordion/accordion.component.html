<h2>Accordion</h2>

<goa-accordion heading="This is the title" data-cy="headingAccordion">
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Animi tempora optio
    illum iste consectetur harum vel voluptatem eos quos deleniti, sequi
    sapiente porro culpa repellendus eligendi in architecto natus omnis. Lorem
    ipsum dolor sit amet consectetur adipisicing elit. Repudiandae, nam
    corporis. Omnis tempora non aspernatur nostrum pariatur quibusdam magni eum
    vitae similique, laboriosam reiciendis repellat sapiente. Nobis laudantium
    facilis incidunt!
  </p>
</goa-accordion>

<goa-accordion heading="This is the title" secondarytext="Secondary text"  data-cy="secondaryAccordion">
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Quidem sapiente
    delectus, hic explicabo doloribus magnam, consequatur eaque veritatis
    possimus porro ducimus vero ipsa omnis, soluta sit voluptates quam facilis
    odio. Lorem, ipsum dolor sit amet consectetur adipisicing elit. Animi,
    officia magnam assumenda nisi exercitationem rerum blanditiis natus modi
    nobis error iure asperiores? Iste vero odio magnam, reprehenderit doloribus
    error iusto.
  </p>
</goa-accordion>

<goa-accordion open="true" headingSize="medium" heading="This is the title" secondarytext="Secondary text"  data-cy="defaultOpenAccordion">
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Quidem sapiente
    delectus, hic explicabo doloribus magnam, consequatur eaque veritatis
    possimus porro ducimus vero ipsa omnis, soluta sit voluptates quam facilis
    odio. Lorem, ipsum dolor sit amet consectetur adipisicing elit. Animi,
    officia magnam assumenda nisi exercitationem rerum blanditiis natus modi
    nobis error iure asperiores? Iste vero odio magnam, reprehenderit doloribus
    error iusto.
  </p>
  <goa-badge type="success" content="Success" slot="headingcontent"></goa-badge>
</goa-accordion>



