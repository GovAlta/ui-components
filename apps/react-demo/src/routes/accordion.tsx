import {
  GoAAccordion,
  GoABadge,
  GoABlock,
  GoAContainer,
  GoAGrid,
} from "@abgov/react-components";
import * as React from "react";

export default function Accordion() {
  return (
    <>
      <GoAAccordion
        heading="This is a heading"
        secondaryText="Secondary Text"
        headingContent={<GoABadge type="success" content="Success" />}
      >
        <a href="#">Some link</a>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Minus dolore
        placeat aperiam officiis, sit corporis nulla dolorem non vero ex nobis
        aliquam cumque amet asperiores. Vitae amet ullam quas nemo?
      </GoAAccordion>
      <GoAAccordion heading="This is a heading" secondaryText="Secondary Text">
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Culpa
        repudiandae ab error exercitationem vero tempora animi odio! Quibusdam
        voluptas laudantium ullam ipsum, fugiat nobis possimus. Cum eum ipsum
        odio ex.
      </GoAAccordion>

      <GoAAccordion
        heading="This is a heading"
        secondaryText="Secondary Text"
        headingContent={
          <div style={{ display: "flex", justifyContent: "flex-end" }}>
            <GoABadge type="success" content="Success" />
          </div>
        }
      >
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Accusantium
        harum aliquid officia provident recusandae facilis dolor minus magnam
        commodi, nesciunt maxime quas aut exercitationem optio modi. Ratione
        commodi maiores aspernatur?
      </GoAAccordion>
      <GoAAccordion heading="Just a large heading" headingSize="medium">
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Nobis,
        similique aut nisi accusamus odio est facere odit dolor sed reiciendis
        quas nostrum dicta quo voluptates, sint illo consectetur laudantium
        blanditiis!
      </GoAAccordion>
      <GoAAccordion
        heading="This is a heading"
        secondaryText="Secondary Text"
        headingContent={
          <div style={{ display: "flex", justifyContent: "flex-end" }}>
            <GoABlock>
              <GoABadge type="success" content="Success" />
              <GoABadge type="emergency" content="Error" />
            </GoABlock>
          </div>
        }
      >
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Accusantium
        harum aliquid officia provident recusandae facilis dolor minus magnam
        commodi, nesciunt maxime quas aut exercitationem optio modi. Ratione
        commodi maiores aspernatur?
      </GoAAccordion>
      <GoAAccordion
        heading="Lorem ipsum dolor sit amet consectetur adipisicing elit. Unde saepe mollitia quo temporibus tempora quos in, perferendis incidunt ex reiciendis praesentium beatae soluta, impedit enim culpa eum quis animi quas!"
        secondaryText="Secondary Text"
        headingContent={
          <div style={{ display: "flex", justifyContent: "flex-end" }}>
            <GoABlock>
              <GoABadge type="success" content="Success" />
              <GoABadge type="emergency" content="Error" />
            </GoABlock>
          </div>
        }
        open
      >
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Culpa
        repudiandae ab error exercitationem vero tempora animi odio! Quibusdam
        voluptas laudantium ullam ipsum, fugiat nobis possimus. Cum eum ipsum
        odio ex.
      </GoAAccordion>
    </>
  );
}
