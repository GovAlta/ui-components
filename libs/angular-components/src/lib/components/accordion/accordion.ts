import { GoabAccordionHeadingSize, Spacing } from "@abgov/ui-components-common";
import { CUSTOM_ELEMENTS_SCHEMA, Component, Input, TemplateRef } from "@angular/core";
import { NgTemplateOutlet } from "@angular/common";

@Component({
  standalone: true,
  selector: "goab-accordion",
  imports: [NgTemplateOutlet],
  template: `
    <goa-accordion
      [attr.heading]="heading"
      [attr.secondarytext]="secondaryText"
      [attr.open]="open"
      [attr.headingsize]="headingSize"
      [attr.maxwidth]="maxWidth"
      [attr.testid]="testId"
      [attr.mt]="mt"
      [attr.mb]="mb"
      [attr.ml]="ml"
      [attr.mr]="mr"
    >
      <div slot="headingcontent">
        <ng-container [ngTemplateOutlet]="headingContent"></ng-container>
      </div>
      <ng-content></ng-content>
    </goa-accordion>
  `,
  schemas: [CUSTOM_ELEMENTS_SCHEMA]
})
export class GoabAccordion {
  @Input() heading?: string;
  @Input() secondaryText?: string;
  @Input() testId?: string;
  @Input() open?: boolean;
  @Input() headingSize?: GoabAccordionHeadingSize;
  @Input() headingContent!: TemplateRef<any>;
  @Input() maxWidth?: string;
  @Input() mt?: Spacing;
  @Input() mb?: Spacing;
  @Input() ml?: Spacing;
  @Input() mr?: Spacing;
}
