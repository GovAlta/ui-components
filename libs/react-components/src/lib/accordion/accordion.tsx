import { ReactNode } from "react";
import { Margins } from "../../common/styling";

export type GoAHeadingSize = "small" | "medium";

interface WCProps extends Margins {
  open?: boolean;
  headingSize?: GoAHeadingSize;
  heading: string;
  secondaryText?: string;
  headingContent?: ReactNode;
  maxwidth?: string;
  testid?: string;
}

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    interface IntrinsicElements {
      "goa-accordion": WCProps & React.HTMLAttributes<HTMLElement>;
    }
  }
}

export interface GoAAccordionProps extends Margins {
  open?: boolean;
  headingSize?: GoAHeadingSize;
  secondaryText?: string;
  heading: string;
  headingContent?: ReactNode;
  maxWidth?: string;
  testid?: string;
  children: ReactNode;
}

export function GoAAccordion({
  open,
  heading,
  headingSize,
  secondaryText,
  headingContent,
  maxWidth,
  testid,
  children,
  mt,
  mr,
  mb,
  ml,
}: GoAAccordionProps): JSX.Element {
  return (
    <goa-accordion
      open={open}
      headingSize={headingSize}
      heading={heading}
      secondaryText={secondaryText}
      maxwidth={maxWidth}
      testid={testid}
      mt={mt}
      mr={mr}
      mb={mb}
      ml={ml}
    >
      {headingContent && <div slot="headingcontent">{headingContent}</div>}
      {children}
    </goa-accordion>
  );
}

export default GoAAccordion;
