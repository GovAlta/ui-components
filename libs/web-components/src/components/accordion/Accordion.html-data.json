{
  "name": "goa-accordion",
  "required": ["heading"],
  "attributes": [
    {
      "name": "heading",
      "description": "Sets the heading text.",
      "type": "string"
    },
    {
      "name": "secondarytext",
      "description": "Sets the secondary text.",
      "type": "string"
    },
    {
      "name": "open",
      "default": "false",
      "description": "Sets the state of the accordion container open or closed.",
      "values": [
        {
          "name": "false",
          "description": "Container is closed."
        },
        {
          "name": "true",
          "description": "Container is opened."
        }
      ],
      "type": "boolean"
    },
    {
      "name": "headingsize",
      "default": "small",
      "description": "Sets the heading size of the accordion container heading.",
      "values": [
        {
          "name": "small",
          "description": "Heading size of the accordion container is small."
        },
        {
          "name": "medium",
          "description": "Heading size of the accordion container is medium."
        }
      ],
      "type": "string"
    },
    {
      "name": "headingcontent",
      "description": "Add components to the accordion container heading such as badges.",
      "type": "slot"
    },
    {
      "name": "maxwidth",
      "description": "Sets the maximum width of the accordion.",
      "type": "string"
    },
    {
      "name": "mt",
      "description": "Top margin",
      "valueSet": "spacing"
    },
    {
      "name": "mr",
      "description": "Right margin",
      "valueSet": "spacing"
    },
    {
      "name": "mb",
      "description": "Bottom margin",
      "valueSet": "spacing"
    },
    {
      "name": "ml",
      "description": "Left margin",
      "valueSet": "spacing"
    }
  ],
  "references": [
    {
      "name": "GoA Accordion",
      "url": "{{siteUrl}}/?path=/docs/components-accordion--basic#accordion"
    }
  ]
}
