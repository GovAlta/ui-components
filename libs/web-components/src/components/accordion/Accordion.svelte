<svelte:options customElement="goa-accordion" />

<!-- Script -->
<script lang="ts">
  import { onMount } from "svelte";
  import { calculateMargin } from "../../common/styling";
  import {
    typeValidator,
    toBoolean,
    validateRequired,
  } from "../../common/utils";
  import type { Spacing } from "../../common/styling";

  // Validator
  const [HeadingSizes, validateHeadingSize] = typeValidator(
    "Accordion heading size",
    ["small", "medium"],
  );

  // Types
  type HeadingSize = (typeof HeadingSizes)[number];

  export let open: string = "false";
  export let heading: string = "";
  export let secondarytext: string = "";
  export let headingsize: HeadingSize = "small";
  export let testid: string = "";
  export let mt: Spacing = null;
  export let mr: Spacing = null;
  export let mb: Spacing = "xs";
  export let ml: Spacing = null;

  // Private

  let _hovering: boolean = false;
  let _titleEl: HTMLElement;
  let _headingContentSlotChildren: Element[] = [];

  // Reactive

  $: isOpen = toBoolean(open);

  // Functions

  onMount(() => {
    validateRequired("GoAAccordion", { heading });
    validateHeadingSize(headingsize);
    _headingContentSlotChildren = getChildren();
  });

  function getChildren(): Element[] {
    if (_titleEl) {
      const slot = _titleEl.querySelector("slot") as HTMLSlotElement;
      if (slot) {
        return slot.assignedElements();
      } else {
        // @ts-expect-error
        return [..._titleEl.children] as Element[]; // unit tests
      }
    }
    return [];
  }
</script>

<!-- HTML -->
<div
  style={calculateMargin(mt, mr, mb, ml)}
  class={`
    goa-accordion
  `}
  data-testid={testid}
>
  <details open={isOpen}>
    <!-- svelte-ignore a11y-no-static-element-interactions -->
    <summary
      class={`container-${headingsize}`}
      on:mouseover={() => (_hovering = true)}
      on:mouseout={() => (_hovering = false)}
      on:focus={() => (_hovering = false)}
      on:blur={() => (_hovering = false)}
    >
      <goa-icon
        type="chevron-forward"
        fillcolor={_hovering
          ? "var(--goa-color-interactive-hover)"
          : "var(--goa-color-interactive-default)"}
      ></goa-icon>
      <div class="title" bind:this={_titleEl}>
        <span
          class="heading heading-{headingsize}"
          data-testid={`${testid}-heading`}>{heading}</span
        >
        <span class="secondary-text">{secondarytext}</span>
        <div
          class="heading-content"
          class:heading-content-top={_headingContentSlotChildren.length}
        >
          <slot name="headingcontent" />
        </div>
      </div>
    </summary>
    <div class="content">
      <slot />
    </div>
  </details>
</div>

<!-- Style -->
<style>
  :host {
    box-sizing: border-box;
    font-family: var(--goa-font-family-sans);
    font-size: var(--goa-font-size-4);
  }

  .goa-accordion,
  .goa-accordion * {
    box-sizing: border-box;
  }

  summary {
    min-height: 3.5rem;
    border-width: var(--goa-border-width-s);
    border-style: solid;
    border-radius: var(--goa-border-radius-m);
    background-color: var(--goa-color-greyscale-100);
    border-color: var(--goa-color-greyscale-200);
    color: var(--goa-color-text-default);
    padding: 0.875rem 1rem 0 0;
    cursor: pointer;
    list-style: none;
    display: flex;
    align-items: flex-start;
    position: relative;
  }

  summary.container-medium {
    min-height: 4rem;
  }

  summary::marker, /* Latest Chrome, Edge, Firefox */
  summary::-webkit-details-marker /* Safari */ {
    display: none;
  }

  summary .title {
    display: flex;
    align-items: center;
    flex: 1;
  }

  .title span {
    padding-bottom: var(--goa-space-3xs, 0);
  }

  summary .heading {
    font: var(--goa-typography-heading-s);
    padding-right: 1rem;
  }

  summary .secondary-text {
    font: var(--goa-typography-body-s);
    line-height: 1.5rem;
    padding-right: 1rem;
  }

  summary .heading-content {
    flex: 1;
  }

  .content {
    border-bottom: var(--goa-border-width-s) solid
      var(--goa-color-greyscale-200);
    border-left: var(--goa-border-width-s) solid var(--goa-color-greyscale-200);
    border-right: var(--goa-border-width-s) solid var(--goa-color-greyscale-200);
    border-bottom-left-radius: var(--goa-border-radius-m);
    border-bottom-right-radius: var(--goa-border-radius-m);
    padding: 1.5rem;
    padding-left: 3.5rem;
    padding-right: 2rem;
  }

  .content :global(::slotted(*:last-child)) {
    margin-bottom: 0 !important;
  }

  summary goa-icon {
    padding: 0.125rem 1rem;
  }

  summary.container-medium goa-icon {
    padding: 0.375rem 1rem;
  }

  details[open] goa-icon {
    transform: rotate(90deg);
  }

  details[open] summary {
    border-bottom-left-radius: var(--goa-border-radius-none);
    border-bottom-right-radius: var(--goa-border-radius-none);
  }

  summary:hover {
    background-color: var(--goa-color-greyscale-200);
  }

  summary:focus,
  summary:active {
    background-color: var(--goa-color-greyscale-100);
    outline: none;
  }

  /* Hack to make outline radius work on Safari */
  summary:active::before,
  summary:focus::before {
    content: "";
    position: absolute;
    top: -1px;
    right: -1px;
    bottom: -1px;
    left: -1px;
    border: var(--goa-border-width-l) solid var(--goa-color-interactive-focus);
    border-radius: 4px;
  }

  /* Sizes */
  summary .heading.heading-medium {
    line-height: 2rem;
    font: var(--goa-typography-heading-m);
  }

  @media (--mobile) {
    summary {
      align-items: flex-start;
    }
    summary .title {
      flex-direction: column;
      align-items: flex-start;
      padding-bottom: 0.875rem;
    }
    summary .title span {
      padding-bottom: 0;
    }
    summary .heading-content.heading-content-top {
      margin-top: var(--goa-space-xs);
    }
  }
</style>
