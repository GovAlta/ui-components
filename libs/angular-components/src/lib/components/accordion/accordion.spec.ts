import { GoabAccordion } from "./accordion";
import { ComponentFixture, TestBed } from "@angular/core/testing";
import { CUSTOM_ELEMENTS_SCHEMA, Component } from "@angular/core";
import { By } from "@angular/platform-browser";
import { GoabAccordionHeadingSize, GoabAccordionIconPosition } from "@abgov/ui-components-common";

@Component({
  template: `
    <goab-accordion
    [heading]="heading"
    [secondaryText]="secondaryText"
    [open]="open"
    [headingSize]="headingSize"
    [headingContent]="headingContent"
    [iconPosition]="iconPosition"
    maxWidth="480px"
  >
    test content
      <ng-template #headingContent>This is the headingcontent</ng-template>
  </goab-accordion>`,
})
class TestAccordionComponent {
  heading?: string;
  secondaryText?: string;
  open?: boolean;
  headingSize?: GoabAccordionHeadingSize;
  iconPosition?: GoabAccordionIconPosition;
}

describe("GoabAccordion", () => {
  let fixture: ComponentFixture<TestAccordionComponent>;
  let component: TestAccordionComponent;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [GoabAccordion],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      declarations: [TestAccordionComponent],
    }).compileComponents();

    fixture = TestBed.createComponent(TestAccordionComponent);
    component = fixture.componentInstance;
    component.heading = "hi";
    component.secondaryText = "Secondary Text";
    component.open = true;
    component.headingSize = "large" as GoabAccordionHeadingSize;
    component.iconPosition = "right" as GoabAccordionIconPosition;

    fixture.detectChanges();
  });

  it("should render and set the props correctly", () => {
    const accordionElement = fixture.debugElement.query(
      By.css("goa-accordion"),
    ).nativeElement;
    expect(accordionElement.getAttribute("heading")).toBe("hi");
    expect(accordionElement.getAttribute("secondarytext")).toBe("Secondary Text");
    expect(accordionElement.getAttribute("open")).toBe("true");
    expect(accordionElement.getAttribute("headingsize")).toBe("large");
    expect(accordionElement.getAttribute("maxwidth")).toBe("480px");
    expect(accordionElement.getAttribute("iconposition")).toBe("right");
    const headingContent = accordionElement.querySelector("[slot='headingcontent']");
    expect(headingContent.textContent).toContain("This is the headingcontent");
  });
});
