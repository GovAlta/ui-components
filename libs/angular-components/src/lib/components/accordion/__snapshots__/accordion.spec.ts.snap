// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<ng-component
  heading={[Function String]}
>
  <goab-accordion>
    <goa-accordion>
      <div
        slot="headingcontent"
      >
         This is the headingcontent
      </div>
       test content
    </goa-accordion>
  </goab-accordion>
</ng-component>
`;
